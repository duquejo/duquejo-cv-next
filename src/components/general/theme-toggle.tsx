'use client';

import { Moon, Sun } from 'lucide-react';
import { useTheme } from 'next-themes';

import { Button, type ButtonProps } from '@/components/ui/button';
import {
  DropdownMenu,
  DropdownMenuContent,
  DropdownMenuItem,
  DropdownMenuTrigger,
} from '@/components/ui/dropdown-menu';
import { cn } from '@/lib/utils';

interface Props {
  className?: string;
  variant?: ButtonProps['variant'];
  title: string;
  light: string;
  dark: string;
  system: string;
}

export function ThemeToggle({ className, variant, title, dark, system, light }: Props) {
  const { setTheme, theme } = useTheme();

  return (
    <DropdownMenu>
      <DropdownMenuTrigger asChild>
        <Button variant={variant} size="icon" className={cn('h-10 w-10', className)} title={title}>
          <Sun
            className="h-[1.2rem] w-[1.2rem] rotate-0 scale-100 transition-all dark:-rotate-90 dark:scale-0"
            aria-label={`${light}-control`}
          />
          <Moon
            className="absolute h-[1.2rem] w-[1.2rem] rotate-90 scale-0 transition-all dark:rotate-0 dark:scale-100"
            aria-label={`${dark}-control`}
          />
          <span className="sr-only">{title}</span>
        </Button>
      </DropdownMenuTrigger>
      <DropdownMenuContent align="end">
        <DropdownMenuItem
          aria-label={light}
          className={cn('text-xs', theme === 'light' && 'bg-accent')}
          onClick={() => setTheme('light')}
        >
          {light}
        </DropdownMenuItem>
        <DropdownMenuItem
          aria-label={dark}
          className={cn('text-xs', theme === 'dark' && 'bg-accent')}
          onClick={() => setTheme('dark')}
        >
          {dark}
        </DropdownMenuItem>
        <DropdownMenuItem
          aria-label={system}
          className={cn('text-xs', theme === 'system' && 'bg-accent')}
          onClick={() => setTheme('system')}
        >
          {system}
        </DropdownMenuItem>
      </DropdownMenuContent>
    </DropdownMenu>
  );
}
