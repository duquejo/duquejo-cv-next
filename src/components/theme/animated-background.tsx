export const AnimatedBackground = () => {
  return (
    <>
      <div className="hidden sm:block fixed z-[1] bottom-0 left-[-20%] right-0 top-[-10%] h-[500px] w-[500px] rounded-full bg-[radial-gradient(circle_farthest-side,rgba(34,197,94,.15),rgba(255,255,255,0))]" />
      <div className="hidden sm:block fixed z-[1] bottom-0 right-[-20%] top-[-10%] h-[500px] w-[500px] rounded-full bg-[radial-gradient(circle_farthest-side,rgba(34,197,94,.15),rgba(255,255,255,0))]" />
    </>
  );
};
