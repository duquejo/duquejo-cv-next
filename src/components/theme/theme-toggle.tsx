'use client';

import { Moon, Sun } from 'lucide-react';
import { useTheme } from 'next-themes';

import { Button, type ButtonProps } from '@/components/ui/button';
import {
  DropdownMenu,
  DropdownMenuContent,
  DropdownMenuItem,
  DropdownMenuTrigger,
} from '@/components/ui/dropdown-menu';
import { cn } from '@/lib';

interface Props {
  className?: string;
  variant?: ButtonProps['variant'];
  title: string;
  light: string;
  dark: string;
  system: string;
}

export function ThemeToggle({ className, variant, title, dark, system, light }: Props) {
  const { setTheme, theme, themes } = useTheme();

  const themeLabels: Record<string, string> = {
    dark,
    light,
    system,
  };

  return (
    <DropdownMenu>
      <DropdownMenuTrigger asChild>
        <Button variant={variant} size="icon" className={cn('h-10 w-10', className)} title={title}>
          <Sun
            className="h-[1.2rem] w-[1.2rem] rotate-0 scale-100 transition-all dark:-rotate-90 dark:scale-0"
            aria-label={light}
          />
          <Moon
            className="absolute h-[1.2rem] w-[1.2rem] rotate-90 scale-0 transition-all dark:rotate-0 dark:scale-100"
            aria-label={dark}
          />
          <span className="sr-only">{title}</span>
        </Button>
      </DropdownMenuTrigger>
      <DropdownMenuContent align="end">
        {themes.map((t: string) => (
          <DropdownMenuItem
            key={t}
            aria-label={themeLabels[t]}
            className={cn('text-xs cursor-pointer', theme === t && 'bg-accent')}
            onClick={() => setTheme(t)}
          >
            {themeLabels[t]}
          </DropdownMenuItem>
        ))}
      </DropdownMenuContent>
    </DropdownMenu>
  );
}
