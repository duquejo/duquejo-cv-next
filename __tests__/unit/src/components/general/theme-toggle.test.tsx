import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { ThemeToggle } from '@/components/theme/theme-toggle';
import userEvent, { type UserEvent } from '@testing-library/user-event';
import { useTheme, type UseThemeProps } from 'next-themes';

vi.mock('next-themes', () => ({
  useTheme: vi.fn(),
}));

describe('<ThemeToggle /> tests', () => {
  const setThemeMock = vi.fn();
  let user: UserEvent;

  const props = {
    title: 'mocked title',
    light: 'Claro',
    dark: 'Oscuro',
    system: 'Sistema',
  };

  const mockThemeArgs: UseThemeProps = {
    theme: 'light',
    themes: ['light', 'dark', 'system'],
    setTheme: setThemeMock,
  };

  beforeEach(() => {
    vi.clearAllMocks();
    vi.mocked(useTheme).mockReturnValue({ ...mockThemeArgs });

    user = userEvent.setup();
  });

  it('Should match the snapshot with the default args', () => {
    const { container } = render(<ThemeToggle {...props} />);

    const button = screen.getByRole('button');

    expect(container).toMatchSnapshot();

    expect(button).toBeInTheDocument();
    expect(button).toHaveAttribute('data-state', 'closed');
    expect(button).toHaveAttribute('aria-expanded', 'false');
    expect(button).toHaveAttribute('title', props.title);
    expect(button).toHaveTextContent(props.title);
    expect(button).toHaveClass('text-primary-foreground', 'hover:bg-primary/90');
  });

  it('Should match the snapshot with a given args', async () => {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { title, ...rest } = props;
    const { container } = render(
      <ThemeToggle variant="secondary" title="Testing toggle" className="bg-color-red" {...rest} />,
    );

    const button = screen.getByRole('button');
    const btnSvgs = container.querySelectorAll('svg');

    expect(container).toMatchSnapshot();

    expect(button).toBeInTheDocument();
    expect(button).toHaveAttribute('title', 'Testing toggle');
    expect(button).toHaveTextContent('Testing toggle');
    expect(button).toHaveClass(
      'text-secondary-foreground',
      'hover:bg-secondary/80',
      'bg-color-red',
    );

    expect(btnSvgs).toHaveLength(2);
  });

  it('Should handle the system theme by default', async () => {
    vi.mocked(useTheme).mockReturnValue({
      ...mockThemeArgs,
      theme: 'system',
      systemTheme: 'dark',
    });

    render(<ThemeToggle variant="secondary" {...props} />);

    expect(screen.getByLabelText(props.light)).toBeInTheDocument();

    await user.click(screen.getByRole('button'));

    await waitFor(() => {
      const selectedThemeOption = screen.getByText(props.system);
      expect(selectedThemeOption).toHaveClass('bg-accent');

      fireEvent.click(selectedThemeOption);
    });

    expect(setThemeMock).toHaveBeenCalledWith('system');
  });

  it('Should handle the light theme', async () => {
    vi.mocked(useTheme).mockReturnValue({ ...mockThemeArgs });

    render(<ThemeToggle {...props} />);

    expect(screen.getByLabelText(props.light)).toBeInTheDocument();

    await user.click(screen.getByRole('button'));

    await waitFor(() => {
      const selectedThemeOption = screen.getByText(props.light);
      expect(selectedThemeOption).toHaveClass('bg-accent');

      fireEvent.click(selectedThemeOption);
    });

    expect(setThemeMock).toHaveBeenCalledWith('light');
  });

  it('Should handle the dark theme', async () => {
    vi.mocked(useTheme).mockReturnValue({
      ...mockThemeArgs,
      theme: 'dark',
      systemTheme: 'dark',
    });

    render(<ThemeToggle {...props} />);

    await user.click(screen.getByRole('button'));

    await waitFor(() => {
      const selectedThemeOption = screen.getByText(props.dark);
      expect(selectedThemeOption).toHaveClass('bg-accent');

      fireEvent.click(selectedThemeOption);
    });

    expect(setThemeMock).toHaveBeenCalledWith('dark');
  });

  it('Should trigger the button click events', async () => {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { title, ...rest } = props;
    render(
      <ThemeToggle variant="secondary" title="Testing toggle" className="bg-color-red" {...rest} />,
    );

    const button = screen.getByRole('button');
    const buttonWrapper = button.parentElement!;

    await user.click(button);

    expect(button).toHaveAttribute('data-state', 'open');
    expect(button).toHaveAttribute('aria-expanded', 'true');
    expect(buttonWrapper).toHaveAttribute('aria-hidden', 'true');
    expect(buttonWrapper).toHaveAttribute('data-aria-hidden', 'true');

    const themeOptionsDropdown = await waitFor(() => screen.getAllByRole('menuitem'));

    expect(themeOptionsDropdown).toHaveLength(3);
  });
});
