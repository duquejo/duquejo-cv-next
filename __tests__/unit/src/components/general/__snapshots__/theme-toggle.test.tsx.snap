// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ThemeToggle /> tests > Should match the snapshot with a given args 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="menu"
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-hidden focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 text-secondary-foreground shadow-xs hover:bg-secondary/80 h-10 w-10 bg-color-red"
    data-state="closed"
    id="radix-:r2:"
    title="Testing toggle"
    type="button"
  >
    <svg
      aria-label="Claro"
      class="lucide lucide-sun h-[1.2rem] w-[1.2rem] rotate-0 scale-100 transition-all dark:-rotate-90 dark:scale-0"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="12"
        cy="12"
        r="4"
      />
      <path
        d="M12 2v2"
      />
      <path
        d="M12 20v2"
      />
      <path
        d="m4.93 4.93 1.41 1.41"
      />
      <path
        d="m17.66 17.66 1.41 1.41"
      />
      <path
        d="M2 12h2"
      />
      <path
        d="M20 12h2"
      />
      <path
        d="m6.34 17.66-1.41 1.41"
      />
      <path
        d="m19.07 4.93-1.41 1.41"
      />
    </svg>
    <svg
      aria-label="Oscuro"
      class="lucide lucide-moon absolute h-[1.2rem] w-[1.2rem] rotate-90 scale-0 transition-all dark:rotate-0 dark:scale-100"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 3a6 6 0 0 0 9 9 9 9 0 1 1-9-9Z"
      />
    </svg>
    <span
      class="sr-only"
    >
      Testing toggle
    </span>
  </button>
</div>
`;

exports[`<ThemeToggle /> tests > Should match the snapshot with the default args 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="menu"
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-hidden focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-primary text-primary-foreground shadow-sm hover:bg-primary/90 h-10 w-10"
    data-state="closed"
    id="radix-:r0:"
    title="mocked title"
    type="button"
  >
    <svg
      aria-label="Claro"
      class="lucide lucide-sun h-[1.2rem] w-[1.2rem] rotate-0 scale-100 transition-all dark:-rotate-90 dark:scale-0"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="12"
        cy="12"
        r="4"
      />
      <path
        d="M12 2v2"
      />
      <path
        d="M12 20v2"
      />
      <path
        d="m4.93 4.93 1.41 1.41"
      />
      <path
        d="m17.66 17.66 1.41 1.41"
      />
      <path
        d="M2 12h2"
      />
      <path
        d="M20 12h2"
      />
      <path
        d="m6.34 17.66-1.41 1.41"
      />
      <path
        d="m19.07 4.93-1.41 1.41"
      />
    </svg>
    <svg
      aria-label="Oscuro"
      class="lucide lucide-moon absolute h-[1.2rem] w-[1.2rem] rotate-90 scale-0 transition-all dark:rotate-0 dark:scale-100"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 3a6 6 0 0 0 9 9 9 9 0 1 1-9-9Z"
      />
    </svg>
    <span
      class="sr-only"
    >
      mocked title
    </span>
  </button>
</div>
`;
